### ISSUE TITLE
Emscripten v1.38.28 ==>  GLSL shader compile fails when using -s USE_PTHREADS=1

### ISSUE THREAD

**Issue by briangavin** [emscripten-core/emscripten#8238]
When I run my app without USE_PTHREADS it works correctly, but when I set -s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=2.   All of my GLSL shaders fail to compile.   When I was running version 1.38.15 pthreads correctly compiled the GLSL shader.   When I switched to version 1.38.28, I get a failed GLSL shader compile error.

I debugged the generated javascript file and found the function UTF8ArrayToString was returning undefined.   Looks like maybe the 'return str' is missing from pthreads js version?

Here is what the 2 versions of the function look like

```
// pthreads off and it works correctly
function UTF8ArrayToString(u8Array, idx, maxBytesToRead) {
            var endIdx = idx + maxBytesToRead;
            var endPtr = idx;
            while (u8Array[endPtr] && !(endPtr >= endIdx))
                ++endPtr;
            if (endPtr - idx > 16 && u8Array.subarray && UTF8Decoder) {
                return UTF8Decoder.decode(u8Array.subarray(idx, endPtr))
            } else {
                var str = "";
                while (idx < endPtr) {
                    var u0 = u8Array[idx++];
                    if (!(u0 & 128)) {
                        str += String.fromCharCode(u0);
                        continue
                    }
                    var u1 = u8Array[idx++] & 63;
                    if ((u0 & 224) == 192) {
                        str += String.fromCharCode((u0 & 31) << 6 | u1);
                        continue
                    }
                    var u2 = u8Array[idx++] & 63;
                    if ((u0 & 240) == 224) {
                        u0 = (u0 & 15) << 12 | u1 << 6 | u2
                    } else {
                        u0 = (u0 & 7) << 18 | u1 << 12 | u2 << 6 | u8Array[idx++] & 63
                    }
                    if (u0 < 65536) {
                        str += String.fromCharCode(u0)
                    } else {
                        var ch = u0 - 65536;
                        str += String.fromCharCode(55296 | ch >> 10, 56320 | ch & 1023)
                    }
                }
            }
            return str
        }   

// pthread on and return undefined string (code `fails)`
function UTF8ArrayToString(u8Array, idx, maxBytesToRead) {
    var endIdx = idx + maxBytesToRead;
    var str = "";
    while (!(idx >= endIdx)) {
        var u0 = u8Array[idx++];
        if (!u0)
            return str;
        if (!(u0 & 128)) {
            str += String.fromCharCode(u0);
            continue
        }
        var u1 = u8Array[idx++] & 63;
        if ((u0 & 224) == 192) {
            str += String.fromCharCode((u0 & 31) << 6 | u1);
            continue
        }
        var u2 = u8Array[idx++] & 63;
        if ((u0 & 240) == 224) {
            u0 = (u0 & 15) << 12 | u1 << 6 | u2
        } else {
            u0 = (u0 & 7) << 18 | u1 << 12 | u2 << 6 | u8Array[idx++] & 63
        }
        if (u0 < 65536) {
            str += String.fromCharCode(u0)
        } else {
            var ch = u0 - 65536;
            str += String.fromCharCode(55296 | ch >> 10, 56320 | ch & 1023)
        }
    }
}

**Comment by briangavin** [comment]
I modified the file runtime_strings.js around line#103 and added the following code
```
#if TEXTDECODER
  }
  return str;
#else
  return str;
#endif // TEXTDECODER

```
to the function and it fixed the problem for my app.

**Comment by kripken** [comment]
Thanks!

The difference in the code is because we disable TEXTDECODER for pthreads (doesn't work with SharedArrayBuffer). Yeah, we didn't handle the case without TEXTDECODER, and when falling out of the loop. Fix in #8252, basically like what you wrote.
